hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/ExpressionTestBase.java

```diff
@@ -0,0 +1,1190 @@
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public abstract class ExpressionTestBase extends SqlTestSupport {
+
+    private static final boolean VERIFY_EVALUATION = true;
+
+    private static final boolean TRACE = true;
+    private static final boolean LOG_ON_SUCCESS = false;
+
+    protected static final RelDataType UNKNOWN_TYPE = null;
+    protected static final SqlTypeName UNKNOWN_TYPE_NAME = null;
+
+    protected static final Object UNKNOWN_VALUE = new Object() {
+        @Override
+        public String toString() {
+            return "UNKNOWN";
+        }
+    };
+
+    protected static final Object INVALID_VALUE = new Object() {
+        @Override
+        public String toString() {
+            return "INVALID";
+        }
+    };
+    protected static final BigDecimal INVALID_NUMERIC_VALUE = new BigDecimal(0);
+    @SuppressWarnings({"checkstyle:IllegalInstantiation", "UnnecessaryBoxing", "BooleanConstructorCall"})
+    protected static final Boolean INVALID_BOOLEAN_VALUE = new Boolean(true);
+
+    public static final HazelcastTypeFactory TYPE_FACTORY = HazelcastTypeFactory.INSTANCE;
+
+    protected static final List<Operand> COLUMNS = new ArrayList<>();
+    protected static final List<Operand> LITERALS = new ArrayList<>();
+    protected static final List<Operand> PARAMETERS = new ArrayList<>();
+    protected static final List<Operand> TYPES = new ArrayList<>();
+
+    protected static final List<Operand> BOOLEAN_COLUMN = new ArrayList<>();
+
+    protected static final SqlOperator IDENTITY = new SqlOperator("", SqlKind.OTHER, 0, 0, null, null, null) {
+        @Override
+        public SqlSyntax getSyntax() {
+            return SqlSyntax.PREFIX;
+        }
+
+        @Override
+        public SqlOperandCountRange getOperandCountRange() {
+            return SqlOperandCountRanges.of(1);
+        }
+
+        @Override
+        public String toString() {
+            return "IDENTITY";
+        }
+    };
+
+    private static final SqlTypeName[] TYPE_NAMES;
+
+    static {
+        TYPE_NAMES = new SqlTypeName[]{VARCHAR, BOOLEAN, TINYINT, SMALLINT, INTEGER, BIGINT, DECIMAL, REAL, DOUBLE, ANY, NULL};
+    }
+
+    private static final Map<String, QueryDataType> FIELDS;
+    private static final Map<String, Integer> FIELD_TO_INDEX;
+    private static final PlanNodeSchema SCHEMA;
+
+    static {
+        Map<String, QueryDataType> fields = new LinkedHashMap<>();
+        for (SqlTypeName type : TYPE_NAMES) {
+            if (type == NULL) {
+                // null is synthetic internal type
+                continue;
+            }
+
+            fields.put(type.getName().toLowerCase() + "1", SqlToQueryType.map(type));
+            fields.put(type.getName().toLowerCase() + "2", SqlToQueryType.map(type));
+        }
+
+        FIELDS = Collections.unmodifiableMap(fields);
```

Comment:
It seems that certain types are not tested: character, big integer

Reply:
Calcite doesn't know anything about non-canonical types and there are no non-canonical types on expressions level.

Reply:
I am not sure why the Calcite is the only concern hew. Consider the `ColumnTest`. It generates queries like this:
```
query:  SELECT varchar1 FROM t
parsed: SELECT `t`.`varchar1`
FROM `t` AS `t`
actual:   VARCHAR CHARACTER SET "UTF-16LE" -> VARCHAR CHARACTER SET "UTF-16LE"
expected: VARCHAR CHARACTER SET "UTF-16LE" -> VARCHAR CHARACTER SET "UTF-16LE"
```
What is the reason that we test, `BigDecimal` and `String`, but `BigInteger` and `Character`? 

In the future, when temporal types are added, we will have 5 types mapped to the `TIMESTAMP_WITH_TIME_ZONE` family. Will they also be excluded from these tests?

Reply:
The tests internally create a mock environment for expression parsing, type inference, validation, conversion and evaluation. Non-canonical types for a certain type family are not making any difference for parsing, type inference, validation and conversion because for Calcite they are represented by a single canonical type. During evaluation, column values are directly passed to expressions by the tests, expressions can't accept values of non-canonical types, to accept values of non-canonical types they must be converted to values of canonical types. So the only thing the addition of non-canonical types would bring is the conversion of values of non-canonical types to canonical ones on the test side before passing them to expressions.